clustering/ejb/infinispan/src/main/java/org/wildfly/clustering/ejb/infinispan/timer/TimerScheduler.java

```diff
@@ -138,45 +138,53 @@ public Boolean call() throws Exception {
+                            InfinispanEjbLogger.ROOT_LOGGER.debugf("Unexpected timeout event triggered for %s", id);
                             return false;
                         }
+
+                        Instant now = Instant.now();
                         Instant currentTimeout = currentTimeoutReference.get();
-                        if (currentTimeout.isAfter(Instant.now())) {
-                            InfinispanEjbLogger.ROOT_LOGGER.debugf("Timeout for timer %s initiated prematurely.", id);
+                        if (currentTimeout.isAfter(now)) {
+                            InfinispanEjbLogger.ROOT_LOGGER.debugf("Timeout for timer %s initiated prematurely @ %s", id, currentTimeout);
                             return false;
                         }
 
-                        Timer<I> timer = factory.createTimer(id, metaData, manager, scheduler);
-
-                        InfinispanEjbLogger.ROOT_LOGGER.debugf("Triggering timeout for timer %s [%s]", id, timer.getMetaData().getContext());
-
-                        // In case we need to reset the last timeout
-                        Optional<Instant> lastTimeout = metaData.getLastTimeout();
-                        // Record last timeout - expected to be set prior to triggering timeout
+                        // Capture previous last timeout in case we need to reset it
+                        Optional<Instant> originalLastTimeout = metaData.getLastTimeout();
+                        // Record new last timeout - expected to be set prior to triggering timeout
                         metaData.setLastTimeout(currentTimeout);
 
-                        try {
-                            timer.invoke();
-                        } catch (ExecutionException e) {
-                            // Log error and proceed as if it was successful
-                            InfinispanEjbLogger.ROOT_LOGGER.error(e.getLocalizedMessage(), e);
-                        } catch (RejectedExecutionException e) {
-                            // Component is not started or is suspended
-                            InfinispanEjbLogger.ROOT_LOGGER.debugf("EJB component is suspended - could not invoke timeout for timer %s", id);
-                            // Reset last timeout
-                            metaData.setLastTimeout(lastTimeout.orElse(null));
-                            return false;
-                        }
-
-                        // If timeout callback canceled this timer
-                        if (timer.isCanceled()) {
-                            InfinispanEjbLogger.ROOT_LOGGER.debugf("Timeout callback canceled timer %s", id);
-                            return true;
-                        }
-
                         // Determine next timeout
                         Optional<Instant> nextTimeout = metaData.getNextTimeout();
+                        // WFLY-19361: If next timeout is in the past do not trigger event
```

Comment:
I have to confess, I don't see how this code works to fire one event to represent a possible sequence of events that may have fired while the timer service is stopped. We have a timer with a getLastTimeout() setting and a getNextTimeout() setting and we have the time now. In line 147, if the next timeout is after now (in the future), don't fire prematurely - agreed. In line 160, if the next timeout is in the past, don't fire anything - not sure how this represents a set of possible firings while the timer service was stopped. In line 163, fire an event if the next firing is exactly equal to the time now - unlikely to be the case. There must be something else going on here, for example in the way getNextTimeout() is calculated? I'm thinking of the case where a single interval timer was created which fires every second, the timer fires once (#1), the timer service is offline for 3 seconds (#2, #3, #4) and then comes back on line, and is now ready to fire again (#5). I'm expecting to see three timeout method invocations - one for #1, one for #2+#3+#4 (to represent those firings which happened when the timer service was offline) and one for #5.   This is why line 160 puzzles me.

Reply:
This works because this logic is only used to determine the _next_ timeout.  Remember that the "last timeout" identifies the last _successful_ timeout.  So, in your example, when the server comes back online, the value for the last timeout will identify execution \#1.  The timer service will schedule the timer to fire per \#2.  When \#2 is successful, then next scheduled event will correspond to \#5.